### PR TITLE
Transformer: add a simple UI wrapper for GPT-2

### PR DESCRIPTION
**Pull request by compnerd**
This adds a simple UI wrapper for the transformer demo.

### CODE REVIEW

File: Transformer/TransformerDemoUI.swift [hunk 1]
```diff
@@ -0,0 +1,241 @@
+import WinSDK
```
Comment (texasmichelle) [inline Transformer/TransformerDemoUI.swift]: I'm getting a compiler error running from the repo root on ubuntu:

```
$ swift run -c release TransformerDemo 0.5 "Introducing Swift for TensorFlow"
.../compnerd/swift-models/Transformer/TransformerDemoUI.swift:1:8: error: no such module 'WinSDK'
import WinSDK
       ^
```

Is this missing from Package.swift or something I need to install in my environment?

File: Transformer/TransformerDemoUI.swift [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      self.gpt = GPT2()
+      self.buttonDelegate.gpt = self.gpt
+      self.label.text = "Transformer ready!"
+    }
+
+    return true
+  }
+}
+
+ApplicationMain(CommandLine.argc, CommandLine.unsafeArgv, nil, SwiftApplicationDelegate())
```
Comment (texasmichelle) [inline Transformer/TransformerDemoUI.swift]: I'm getting a compiler error running from the repo root on ubuntu:

```
$ swift run -c release TransformerDemo 0.5 "Introducing Swift for TensorFlow"
.../compnerd/swift-models/Transformer/TransformerDemoUI.swift:241:1: error: expressions are not allowed at the top level
ApplicationMain(CommandLine.argc, CommandLine.unsafeArgv, nil, SwiftApplicationDelegate())
^
```

File: Transformer/TransformerDemoUI.swift [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+import SwiftWin32
+
+import Dispatch
+import Foundation
+import ModelSupport
+import TextModels
+import TensorFlow
+import Transformer
+
+internal extension BytePairEncoder {
```
Comment (texasmichelle) [inline Transformer/TransformerDemoUI.swift]: I think this is broadly applicable enough to be added to Models/Text/BytePairEncoding.swift. It could simplify some of the usage in Models/Text/BERT.swift.

File: Transformer/TransformerDemoUI.swift (hunk 1 above)
Comment (compnerd) [reply Transformer/TransformerDemoUI.swift]: The dependency for that is something which is dependent on Windows itself.  I think we need to figure out how to exclude this from Package.swift by default.

File: Transformer/TransformerDemoUI.swift (hunk 2 above)
Comment (compnerd) [reply Transformer/TransformerDemoUI.swift]: Ah, right, because it must be named `main.swift`.  Adjusted along with namespacing.

File: Transformer/TransformerDemoUI.swift (hunk 1 above)
Comment (dan-zheng) [reply Transformer/TransformerDemoUI.swift]: > I think we need to figure out how to exclude this from Package.swift by default.

I wonder if this requires adding imperative code to `Package.swift`?

Comment (compnerd) [reply Transformer/TransformerDemoUI.swift]: I *think* that `.windows` was added as a platform, so it should be possible to at least mark the file to not be built.

**Comment by Shashi456** [comment]
@compnerd could you put a screenshot in, excited to look at it xD.

File: Transformer/main.swift [hunk 4]
```diff
@@ -12,97 +12,20 @@
...
-let start_token = tokenToID["<|endoftext|>"]!
-var tokens = Tensor(shape: [1, 1], scalars: [start_token])
-var temperature = Float(1.0)
+let gpt: GPT2 = GPT2()!
 
 // Set temperature.
-if CommandLine.arguments.count >= 2 {
-    temperature = Float(CommandLine.arguments[1])!
-}
+gpt.temperature = CommandLine.arguments.count > 2
```
Comment (texasmichelle) [inline Transformer/main.swift]: Should this be `>= 2`? It's possible to include a temperature without seed.

Comment (compnerd) [reply Transformer/main.swift]: Ugh ... clearly, I copy paste good :-(

File: Transformer/TransformerDemoUI.swift (hunk 3 above)
Comment (compnerd) [reply Transformer/TransformerDemoUI.swift]: Moved.

File: Transformer/TransformerLoop.swift [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    seed = Tensor(shape: [1, 1], scalars: [mapping["<|endoftext|>"]!])
+    let empty: Tensor<Float> =
+        Tensor<Float>(zeros: [parameters.headCount, 0,
+                              parameters.embeddingSize / parameters.headCount])
+    states = (0 ..< parameters.layerCount).map { _ in
+      AttentionContext(key: empty, value: empty)
+    }
+  }
+
+  func tensor(for string: String) -> Tensor<Int32> {
```
Comment (compnerd) [inline Transformer/TransformerLoop.swift]: I think that this should be called `embedding`

File: Transformer/TransformerLoop.swift [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        result.slice(lowerBounds: [0, timesteps - 1, 0],
+                     upperBounds: [batchSize, timesteps, vocabularySize]) / temperature
+    seed = Tensor(randomCategorialLogits: logits.squeezingShape(at: 1),
+                  sampleCount: 1)
+
+    let ID: Int32 = Int32(seed[0][0])!
+    if let token: String = mapping.key(ID) {
+      return BytePairEncoder.decode(token: token)
+    }
+    print("ID \(ID) not found!")
```
Comment (compnerd) [inline Transformer/TransformerLoop.swift]: Since this is actually not supposed to happen, perhaps it makes sense to have this function be a `throws` function and return a `String` instead?

Comment (texasmichelle) [reply Transformer/TransformerLoop.swift]: SGTM

File: Transformer/TransformerLoop.swift (hunk 5 above)
Comment (texasmichelle) [reply Transformer/TransformerLoop.swift]: yes, good call!
